src/Hipace.cpp

```diff
@@ -507,39 +504,92 @@ Hipace::SolveOneSlice (int islice, const int islice_local, int step)
     // reorder plasma before TileSort
     m_multi_plasma.ReorderParticles(islice);
 
+    // prepare/initialize fields
     for (int lev=0; lev<current_N_level; ++lev) {
+        m_fields.InitializeSlices(lev, islice, m_3D_geom);
+    }
 
-        if (lev != 0 && islice == m_3D_geom[lev].Domain().bigEnd(Direction::z)) {
-            // first slice of lev (islice goes backwards)
-            // iterpolate jx_beam and jy_beam from lev-1 to lev
-            m_fields.LevelUp(m_3D_geom, lev, WhichSlice::Previous1, "jx_beam");
-            m_fields.LevelUp(m_3D_geom, lev, WhichSlice::Previous1, "jy_beam");
-            m_fields.LevelUp(m_3D_geom, lev, WhichSlice::This, "jx_beam");
-            m_fields.LevelUp(m_3D_geom, lev, WhichSlice::This, "jy_beam");
-            m_fields.duplicate(lev, WhichSlice::This, {"jx"     , "jy"     },
-                                    WhichSlice::This, {"jx_beam", "jy_beam"});
-        }
-
+    // deposit current
+    for (int lev=0; lev<current_N_level; ++lev) {
+        // tiling used by plasma current deposition
         if (m_do_tiling) m_multi_plasma.TileSort(m_slice_geom[lev].Domain(), m_slice_geom[lev]);
 
         if (m_explicit) {
-            ExplicitSolveOneSubSlice(lev, step, islice, islice_local);
+            // deposit jx, jy, chi and rhomjz for all plasmas
+            m_multi_plasma.DepositCurrent(m_fields, m_multi_laser, WhichSlice::This,
+                true, false, m_deposit_rho, true, true, m_3D_geom, lev);
+
+            // deposit jx_beam and jy_beam in the Next slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                m_do_beam_jx_jy_deposition, false, false, WhichSlice::Next);
+
+            // deposit jz_beam and maybe rhomjz of the beam on This slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                false, true, m_do_beam_jz_minus_rho, WhichSlice::This);
         } else {
-            PredictorCorrectorSolveOneSubSlice(lev, step, islice, islice_local);
+            // deposit jx jy jz (maybe chi) and rhomjz
+            m_multi_plasma.DepositCurrent(m_fields, m_multi_laser, WhichSlice::This,
+                true, true, m_deposit_rho, m_use_laser, true, m_3D_geom, lev);
+
+            // deposit jx jy jz and maybe rhomjz on This slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                m_do_beam_jx_jy_deposition, true, m_do_beam_jz_minus_rho, WhichSlice::This);
         }
+        // add neutralizing background
+        m_fields.AddRhoIons(lev);
+
+        // deposit grid current into jz_beam
+        m_grid_current.DepositCurrentSlice(m_fields, m_3D_geom[lev], lev, islice);
+    }
+
+    // Psi Ez Bz solve
```

Comment:
```suggestion
    // Psi ExmBy EypBx Ez Bz solve
```
just for consistency